internal/buildjob/buildjob.go

```diff
@@ -0,0 +1,296 @@
+	"github.com/dominodatalab/forge/internal/message"
+	"github.com/dominodatalab/forge/plugins/preparer"
+
+	_ "github.com/dominodatalab/forge/internal/unshare"
+)
+
+type Job struct {
+	log logr.Logger
+
+	clientk8s   kubernetes.Interface
+	clientforge clientv1alpha1.Interface
+
+	producer message.Producer
+
+	plugins []*preparer.Plugin
+
+	builder builder.OCIImageBuilder
+
+	name      string
+	namespace string
+
+	cleanupSteps []func()
+}
+
+func New(cfg Config) (*Job, error) {
+	log := NewLogger()
+
+	// unshare user, mount namespace and re-exec ourselves
+	reexec(log)
+
+	// initialize kubernetes clients
+	log.Info("Initializing Kubernetes clients")
+
+	restCfg, err := loadKubernetesConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot load k8s config")
+	}
+	clientsk8s, err := kubernetes.NewForConfig(restCfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create k8s api client")
+	}
+	clientforge, err := clientv1alpha1.NewForConfig(restCfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create forge api client")
+	}
+
+	var cleanupSteps []func()
+
+	// setup message publisher
+	var producer message.Producer
+	if cfg.BrokerOpts != nil {
+		log.Info("Initializing status update message publisher")
+
+		if producer, err = message.NewProducer(cfg.BrokerOpts); err != nil {
+			return nil, err
+		}
+		cleanupSteps = append(cleanupSteps, func() {
+			log.Info("Closing message producer")
+			producer.Close()
+		})
+	}
+
+	// setup preparer plugins
+	log.Info("Loading configured preparer plugins")
+
+	preparerPlugins, err := preparer.LoadPlugins(cfg.PreparerPluginsPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to load preparer plugins path %q", cfg.PreparerPluginsPath)
+	}
+	cleanupSteps = append(cleanupSteps, func() {
+		log.Info("Killing preparer plugins")
+		for _, preparerPlugin := range preparerPlugins {
+			preparerPlugin.Kill()
+		}
+	})
+
+	// instantiate the image builder
+	log.Info("Initializing OCI image builder")
+
+	ociBuilder, err := builder.New(preparerPlugins, cfg.EnableLayerCaching, log)
+	if err != nil {
+		return nil, errors.Wrap(err, "image builder initialization failed")
+	}
+
+	return &Job{
+		log:          log,
+		name:         cfg.ResourceName,
+		namespace:    cfg.ResourceNamespace,
+		clientk8s:    clientsk8s,
+		clientforge:  clientforge,
+		producer:     producer,
+		plugins:      preparerPlugins,
+		builder:      ociBuilder,
+		cleanupSteps: cleanupSteps,
+	}, nil
+}
+
+func (j *Job) Run() error {
+	ctx := context.TODO()
+
+	j.log.Info("Fetching ContainerImageBuild resource", "Name", j.name, "Namespace", j.namespace)
+	cib, err := j.clientforge.ContainerImageBuilds(j.namespace).Get(j.name)
+	if err != nil {
+		return errors.Wrapf(err, "cannot find containerimagebuild %s", j.name)
+	}
+
+	if cib, err = j.transitionToBuilding(cib); err != nil {
+		return err
+	}
+
+	j.log = j.log.WithValues("annotations", cib.Annotations)
```

Comment:
Fwiw I think we can simplify some of the fluentd config since now it can just pick up the annotation, but totally fine as a follow-up

Reply:
I _think_ I understand what you're saying but let's chat some more about it please.